apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -0,0 +1,522 @@
+            if let Some(service_namespace) = &config.service_namespace {
+                resource_defaults.push(KeyValue::new(
+                    opentelemetry_semantic_conventions::resource::SERVICE_NAMESPACE,
+                    service_namespace.clone(),
+                ));
+            }
+            let resource = Resource::new(resource_defaults).merge(&mut Resource::new(
+                config
+                    .attributes
+                    .clone()
+                    .unwrap_or_default()
+                    .iter()
+                    .map(|(k, v)| {
+                        KeyValue::new(
+                            opentelemetry::Key::from(k.clone()),
+                            opentelemetry::Value::from(v.clone()),
+                        )
+                    })
+                    .collect::<Vec<KeyValue>>(),
+            ));
+
+            trace_config = trace_config.with_resource(resource);
+            trace_config
+        }
+    }
+}
+
+fn parent_based(sampler: opentelemetry::sdk::trace::Sampler) -> opentelemetry::sdk::trace::Sampler {
+    opentelemetry::sdk::trace::Sampler::ParentBased(Box::new(sampler))
+}
+
+#[derive(Debug)]
+struct ReportingError;
+
+impl fmt::Display for ReportingError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "ReportingError")
+    }
+}
+
+impl std::error::Error for ReportingError {}
+
+trait TracingConfigurator {
+    fn apply(&self, builder: Builder, trace_config: &Trace) -> Result<Builder, BoxError>;
+}
+
+trait GenericWith<T>
+where
+    Self: Sized,
+{
+    fn with<B>(self, option: &Option<B>, apply: fn(Self, &B) -> Self) -> Self {
+        if let Some(option) = option {
+            return apply(self, option);
+        }
+        self
+    }
+    fn try_with<B>(
+        self,
+        option: &Option<B>,
+        apply: fn(Self, &B) -> Result<Self, BoxError>,
+    ) -> Result<Self, BoxError> {
+        if let Some(option) = option {
+            return apply(self, option);
+        }
+        Ok(self)
+    }
+}
+
+impl<T> GenericWith<T> for T where Self: Sized {}
+
+fn setup<T: TracingConfigurator>(
+    mut builder: Builder,
+    configurator: &Option<T>,
+    tracing_config: &Trace,
+) -> Result<Builder, BoxError> {
+    if let Some(config) = configurator {
+        builder = config.apply(builder, tracing_config)?;
+    }
+    Ok(builder)
+}
+
+fn apollo_key() -> Option<String> {
+    std::env::var("APOLLO_KEY").ok()
+}
+
+fn apollo_graph_reference() -> Option<String> {
+    std::env::var("APOLLO_GRAPH_REF").ok()
+}
+
+#[async_trait::async_trait]
+impl Plugin for Telemetry {
+    type Config = config::Conf;
+
+    async fn startup(&mut self) -> Result<(), BoxError> {
+        // Apollo config is special because we enable tracing if some env variables are present.
+        let apollo = self.config.apollo.get_or_insert_with(Default::default);
+        if apollo.apollo_key.is_none() {
+            apollo.apollo_key = apollo_key()
```

Comment:
that should be in the Default implementation for Conf (along with serde's default annotation)

Reply:
This causes the env variable to be output as part of the JSON schema. I'll add a comment as to why this is not done in the default impl.

Reply:
ah, got it